File: crates/deepsec-scanner/src/gate.rs
use crate::detect::DetectedTech;
use crate::matchers::MatcherGate;
use globset::{Glob, GlobSetBuilder};
use std::collections::HashSet;
use std::path::Path;

/// Evaluate a matcher's `requires` gate. Returns true when the matcher
/// should run. Semantics mirror the TS implementation:
///   - empty gate → always runs.
///   - `tech` set → at least one tag must be in `detected.tags`.
///   - `sentinel_files` set → at least one path/glob must exist under
///     `root`. When `sentinel_contains` is set, the file content must
///     also match at least one of those regexes.
///   - Both present → EITHER passing is enough (union, not intersection).
pub fn evaluate_gate(gate: &MatcherGate, detected: &DetectedTech, root: &Path) -> bool {
    if gate.tech.is_empty() && gate.sentinel_files.is_empty() {
        return true;
    }
    if !gate.tech.is_empty() {
        let have: HashSet<&str> = detected.tags.iter().map(String::as_str).collect();
        if gate.tech.iter().any(|t| have.contains(t.as_str())) {
            return true;
        }
    }
    if !gate.sentinel_files.is_empty() {
        if has_sentinel(&gate.sentinel_files, &gate.sentinel_contains, root) {
            return true;
        }
    }
    false
}

fn has_sentinel(patterns: &[String], contains: &[String], root: &Path) -> bool {
    let mut builder = GlobSetBuilder::new();
    let mut any = false;
    for p in patterns {
        if !p.contains('*') {
            // literal path
            if root.join(p).exists() {
                if contains.is_empty() {
                    return true;
                }
                if let Ok(body) = fs_err::read_to_string(root.join(p)) {
                    if contains_any(&body, contains) {
                        return true;
                    }
                }
            }
            continue;
        }
        if let Ok(g) = Glob::new(p) {
            builder.add(g);
            any = true;
        }
    }
    if !any {
        return false;
    }
    let Ok(set) = builder.build() else {
        return false;
    };

    for entry in walkdir::WalkDir::new(root)
        .max_depth(8)
        .into_iter()
        .filter_map(|e| e.ok())
    {
        if !entry.file_type().is_file() {
            continue;
        }
        let rel = match entry.path().strip_prefix(root) {
            Ok(p) => p,
            Err(_) => continue,
        };
        if set.is_match(rel) {
            if contains.is_empty() {
                return true;
            }
            if let Ok(body) = fs_err::read_to_string(entry.path()) {
                if contains_any(&body, contains) {
                    return true;
                }
            }
        }
    }
    false
}

fn contains_any(body: &str, patterns: &[String]) -> bool {
    for p in patterns {
        if let Ok(re) = fancy_regex::Regex::new(p) {
            if re.is_match(body).unwrap_or(false) {
                return true;
            }
        }
    }
    false
}

File: internal/scanner/gate.go
package scanner

import (
	"os"
	"path/filepath"
	"regexp"

	"github.com/bmatcuk/doublestar/v4"
)

// EvaluateGate returns true when the matcher should run against the
// detected tech / root combo. Semantics: empty gate always runs; tech
// and sentinel_files are OR'd (union), not AND'd.
func EvaluateGate(g MatcherGate, tech DetectedTech, root string) bool {
	if len(g.Tech) == 0 && len(g.SentinelFiles) == 0 {
		return true
	}
	if len(g.Tech) > 0 {
		have := map[string]struct{}{}
		for _, t := range tech.Tags {
			have[t] = struct{}{}
		}
		for _, t := range g.Tech {
			if _, ok := have[t]; ok {
				return true
			}
		}
	}
	if len(g.SentinelFiles) > 0 && hasSentinel(g.SentinelFiles, g.SentinelContains, root) {
		return true
	}
	return false
}

func hasSentinel(patterns, contains []string, root string) bool {
	for _, p := range patterns {
		// Literal path? fast-path.
		if !containsAny(p, "*?[") {
			abs := filepath.Join(root, p)
			if _, err := os.Stat(abs); err != nil {
				continue
			}
			if len(contains) == 0 {
				return true
			}
			body, err := os.ReadFile(abs)
			if err == nil && matchesAny(string(body), contains) {
				return true
			}
			continue
		}
		// Glob path.
		match, err := globFind(root, p)
		if err != nil || match == "" {
			continue
		}
		if len(contains) == 0 {
			return true
		}
		body, err := os.ReadFile(match)
		if err == nil && matchesAny(string(body), contains) {
			return true
		}
	}
	return false
}

func globFind(root, pattern string) (string, error) {
	abs, err := doublestar.Glob(os.DirFS(root), pattern)
	if err != nil || len(abs) == 0 {
		return "", err
	}
	return filepath.Join(root, abs[0]), nil
}

func containsAny(s, chars string) bool {
	for _, c := range chars {
		for _, r := range s {
			if r == c {
				return true
			}
		}
	}
	return false
}

func matchesAny(body string, patterns []string) bool {
	for _, p := range patterns {
		r, err := regexp.Compile(p)
		if err != nil {
			continue
		}
		if r.MatchString(body) {
			return true
		}
	}
	return false
}
